File: core-sample-app/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/core/sampleapp/examples/chromecastExample/utils/MediaRouteButtonUtils.java
package com.pierfrancescosoffritti.androidyoutubeplayer.core.sampleapp.examples.chromecastExample.utils;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.util.Log;

import com.google.android.gms.cast.framework.CastButtonFactory;
import com.pierfrancescosoffritti.androidyoutubeplayer.core.sampleapp.examples.chromecastExample.ChromeCastExampleActivity;
import com.pierfrancescosoffritti.aytplayersample.R;

import androidx.annotation.Nullable;
import androidx.appcompat.view.ContextThemeWrapper;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.mediarouter.app.MediaRouteButton;

public class MediaRouteButtonUtils {
    static public MediaRouteButton initMediaRouteButton(Context context) {
        MediaRouteButton mediaRouteButton = new MediaRouteButton(context);
        CastButtonFactory.setUpMediaRouteButton(context, mediaRouteButton);

        return mediaRouteButton;
    }

    static public void addMediaRouteButtonToPlayerUi(
            MediaRouteButton mediaRouteButton, int tintColor,
            @Nullable ChromeCastExampleActivity.MediaRouteButtonContainer disabledContainer, ChromeCastExampleActivity.MediaRouteButtonContainer activatedContainer) {

        setMediaRouterButtonTint(mediaRouteButton, tintColor);

        if(disabledContainer != null)
        disabledContainer.removeMediaRouteButton(mediaRouteButton);

        if(mediaRouteButton.getParent() != null)
            return;

        activatedContainer.addMediaRouteButton(mediaRouteButton);
    }

    static private void setMediaRouterButtonTint(MediaRouteButton mediaRouterButton, int color) {
        ContextThemeWrapper castContext = new ContextThemeWrapper(mediaRouterButton.getContext(), R.style.Theme_MediaRouter);
        TypedArray styledAttributes = castContext.obtainStyledAttributes(null, androidx.mediarouter.R.styleable.MediaRouteButton, androidx.mediarouter.R.attr.mediaRouteButtonStyle, 0);
        Drawable drawable = styledAttributes.getDrawable(androidx.mediarouter.R.styleable.MediaRouteButton_externalRouteEnabledDrawable);

        if(drawable == null) {
            Log.e("MediaRouteButtonUtils", "can't apply tint to MediaRouteButton");
            return;
        }

        styledAttributes.recycle();
        DrawableCompat.setTint(drawable, ContextCompat.getColor(mediaRouterButton.getContext(), color));

        mediaRouterButton.setRemoteIndicatorDrawable(drawable);
    }
}

File: chromecast-sender-sample-app/src/main/java/com/pierfrancescosoffritti/androidyoutubeplayer/chromecast/sampleapp/utils/MediaRouteButtonUtils.kt
package com.pierfrancescosoffritti.androidyoutubeplayer.chromecast.sampleapp.utils

import android.content.Context
import android.view.ContextThemeWrapper
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.mediarouter.R
import androidx.mediarouter.app.MediaRouteButton
import com.google.android.gms.cast.framework.CastButtonFactory
import com.pierfrancescosoffritti.androidyoutubeplayer.chromecast.sampleapp.examples.localPlayerExample.MediaRouteButtonContainer

object MediaRouteButtonUtils {

    fun initMediaRouteButton(context: Context) : MediaRouteButton {
        val mediaRouteButton = MediaRouteButton(context)
        CastButtonFactory.setUpMediaRouteButton(context, mediaRouteButton)

        return mediaRouteButton
    }

    fun initMediaRouteButton(mediaRouteButton: MediaRouteButton) : MediaRouteButton {
        CastButtonFactory.setUpMediaRouteButton(mediaRouteButton.context, mediaRouteButton)
        return mediaRouteButton
    }

    fun addMediaRouteButtonToPlayerUi(
            mediaRouteButton: MediaRouteButton, tintColor: Int,
            disabledContainer: MediaRouteButtonContainer?, activatedContainer: MediaRouteButtonContainer) {

        setMediaRouterButtonTint(mediaRouteButton, tintColor)

        disabledContainer?.removeMediaRouteButton(mediaRouteButton)
        if(mediaRouteButton.parent != null) return
        activatedContainer.addMediaRouteButton(mediaRouteButton)
    }

    private fun setMediaRouterButtonTint(mediaRouterButton: MediaRouteButton, color: Int) {
        val castContext = ContextThemeWrapper(mediaRouterButton.context, R.style.Theme_MediaRouter)
        val styledAttributes = castContext.obtainStyledAttributes(null, R.styleable.MediaRouteButton, R.attr.mediaRouteButtonStyle, 0)
        val drawable = styledAttributes.getDrawable(R.styleable.MediaRouteButton_externalRouteEnabledDrawable)

        styledAttributes.recycle()
        drawable?.let { DrawableCompat.setTint(it, ContextCompat.getColor(mediaRouterButton.context, color)) }

        mediaRouterButton.setRemoteIndicatorDrawable(drawable)
    }
}